### PR TITLE
[Order Details] Add tracking: Implement "Discard changes" ActionSheet

### PR DESCRIPTION
**Pull request by iamgabrielma**
Partial fix for: https://github.com/woocommerce/woocommerce-ios/issues/1949 in conjunction with https://github.com/woocommerce/woocommerce-ios/pull/7148

## Description
We want to avoid users dismissing views by accident, without saving data. This PR implements a "discard changes" alert when a user attempts to dismiss the "Add Tracking" view without saving changes first, either by tapping on the "dismiss" button or by attempting to dismiss the view via the pull-down-to-dismiss gesture.

It adds two new computed variables to the model, which, along with the existing `canCommit`, covers all the cases we need to check if there are unsaved changes.
- `isEmptyState`: All fields are empty, these vary depending on if the shipment provider is custom or not.
- `hasUnsavedChanges`: Some fields are not-empty, these vary depending on if the shipment provider is custom or not.

## Testing instructions:
- Install and activate https://woocommerce.com/products/shipment-tracking/
- All the cases can be tested by going to Orders > Select an Order > Scroll to `(+) Add tracking` :

Simulator Screen Shot - iPhone 11 Pro Max - 2022-06-29 at 22 18 06

### Case 1: Tap dismiss or pull-to-dismiss with empty fields
1. Do not fill Shipping Carrier or Tracking number. Both fields have to be empty.
2. Tap dismiss, or pull-to-dismiss 
3. See how the view is dismissed

### Case 2: Tap dismiss or pull-to-dismiss with unsaved data in 1 field
1. Fill one field: Shipping Carrier or Tracking number
2. Tap dismiss, or pull-to-dismiss 
3. See the Discard Changes alert

### Case 3:  Tap dismiss or pull-to-dismiss with unsaved data in 2 fields
1. Fill both fields: Shipping Carrier and Tracking number
2. Tap dismiss, or pull-to-dismiss 
3. See the Discard Changes alert

### Case 4: Tap Add
1. Fill both fields: Shipping Carrier and Tracking number
2. Tap Add
3. See how the view is dismissed

### Case 5: Custom Carriers
1. Go to Orders > Select an Order > Scroll to `(+) Add tracking`, and in `Shipping Carrier `select `Custom Carrier`
2. Try to tap dismiss or pull-to-dismiss when any of the mandatory fields are empty or we have not saved the data
3. See the Discard Changes alert 

TODO
- [x] Case: Show ActionSheet when the user taps "Dismiss" and changes can't be committed 479c3f2d40e66093318f98cd9bc9b169a080bf17
- [x] Case: Show ActionSheet when the user taps "Dismiss", changes can be committed, but they aren't yet -> Added tracked numbers but haven't tapped "add" 3692817a17fbc8d0cee8e68105440fa29e8f0ad4
- [x] Case: Show ActionSheet when the user attempts to "pull to dismiss" d90d0c95f69e44fdd8f9f8f335df89add3c82ef2
- [x] Edge case: pull-to-dismiss & all fields empty
- [x] Test for Custom providers
- [x] Move checks to the viewModel
- [x] Clean-up logic for pull-to-dismiss

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift [hunk 1]
```diff
@@ -70,6 +72,9 @@ private extension ManualTrackingViewController {
         configureTitle()
         configureDismissButton()
         configureAddButton()
+        // Disables the ability to dismiss the view controller via a pull-down gesture, in order to avoid losing unsaved changes.
+        isModalInPresentation = false
```
Comment (toupper) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift]: Is this really necessary? I see that in other places in our app where we have the same use case we do not have to set this property

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift [hunk 2]
```diff
@@ -34,6 +34,8 @@ final class ManualTrackingViewController: UIViewController {
 
     private var valueSubscriptions: Set<AnyCancellable> = []
 
+    var hasUnsavedChanges: Bool = true
```
Comment (toupper) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift]: I think it is more appropriate if we move this to the view model. Furthermore, It might be less error-prone if we check whether there are changes dynamically, e.g making it a computed property.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift [hunk 3]
```diff
@@ -134,12 +139,17 @@ private extension ManualTrackingViewController {
     }
 
     @objc func dismissButtonTapped() {
-        dismiss()
+        if viewModel.canCommit == false || hasUnsavedChanges {
```
Comment (toupper) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift]: I think we don't care whether the view model can commit or not when dismissing the view, we only care about whether there are unsaved changes or not.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift [hunk 4]
```diff
@@ -448,6 +458,20 @@ private extension ManualTrackingViewController {
     }
 }
 
+// MARK: - UISheetPresentationControllerDelegate comformance
+//
+/// Asks permission to dismiss. We call this delegate method whenever the user attempts to dismiss the View Controller via the pull-down gesture.
+extension ManualTrackingViewController: UISheetPresentationControllerDelegate {
+    func presentationControllerShouldDismiss(_ presentationController: UIPresentationController) -> Bool {
+        if hasUnsavedChanges {
```
Comment (toupper) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift]: I guess we can move the actual showing the confirmation alert to `presentationControllerDidAttemptToDismiss`, leaving in `presentationControllerShouldDismiss` just the logic of whether the should dismiss or not.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift [hunk 5]
```diff
@@ -547,6 +571,12 @@ private extension ManualTrackingViewController {
         ServiceLocator.stores.dispatch(action)
 
     }
+    func displayDismissConfirmationAlert() {
+        UIAlertController.presentDiscardChangesActionSheet(viewController: self,
+                                                           onDiscard: {[weak self] in self?.dismiss(animated: true)},
+                                                           onCancel: .none
```
Comment (toupper) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift]: Do we need the `onCancel` parameter here? By default is nil, so I guess we can remove it altogether.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift (hunk 1 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift]: It is not necessary indeed 👍

File: WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift [hunk 6]
```diff
@@ -128,6 +129,12 @@ final class AddTrackingViewModel: ManualTrackingViewModel {
     }
 
     let providerCellAccessoryType = UITableViewCell.AccessoryType.disclosureIndicator
+    var isEmptyState: Bool {
+        (shipmentProvider == nil && trackingNumber.isNilOrEmpty) ? true : false
```
Comment (toupper) [inline WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift]: You can add directly`(shipmentProvider == nil && trackingNumber.isNilOrEmpty)`,without the need of adding the ternary operator` ? true : false`

File: WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift [hunk 7]
```diff
@@ -128,6 +129,12 @@ final class AddTrackingViewModel: ManualTrackingViewModel {
...
+    }
```
Comment (toupper) [inline WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift]: Super not a blocker, you can leave an empty line between computer properties

File: WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift [hunk 8]
```diff
@@ -128,6 +129,12 @@ final class AddTrackingViewModel: ManualTrackingViewModel {
...
+    var hasUnsavedChanges: Bool {
+        (shipmentProvider == nil || trackingNumber.isNilOrEmpty) ? true : false
```
Comment (toupper) [inline WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift]: same here about the ternary operator

File: WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift [hunk 9]
```diff
@@ -231,6 +238,12 @@ final class AddCustomTrackingViewModel: ManualTrackingViewModel {
     }
 
     let providerCellAccessoryType = UITableViewCell.AccessoryType.none
+    var isEmptyState: Bool {
```
Comment (toupper) [inline WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift]: We have this code duplicated above, is there a way to avoid duplicating this? Perhaps through a protocol extension? some utility class might do this work?

File: WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift [hunk 10]
```diff
@@ -231,6 +238,12 @@ final class AddCustomTrackingViewModel: ManualTrackingViewModel {
...
+        (providerName.isNilOrEmpty && trackingNumber.isNilOrEmpty) ? true : false
+    }
+    var hasUnsavedChanges: Bool {
```
Comment (toupper) [inline WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift:236]: If `providerName.isNilOrEmpty` is `true` and `trackingNumber.isNilOrEmpty` is also `true`, we would return `hasUnsavedChanges` to `true`, and I don't think that is correct. Am I right?

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift [hunk 11]
```diff
@@ -134,7 +136,13 @@ private extension ManualTrackingViewController {
     }
 
     @objc func dismissButtonTapped() {
-        dismiss()
+        if viewModel.canCommit || viewModel.hasUnsavedChanges {
```
Comment (toupper) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift]: Do we need to check both `viewModel.canCommit || viewModel.hasUnsavedChanges` here, or `viewModel.hasUnsavedChanges` would be enough?

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift [hunk 12]
```diff
@@ -134,7 +136,13 @@ private extension ManualTrackingViewController {
...
+            displayDismissConfirmationAlert()
+        } else if viewModel.isEmptyState {
```
Comment (toupper) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift]: It is not necessary to add the `else if viewModel.isEmptyState {` here, with just 
```
} else {
            dismiss()
        }
```
it is enough.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift [hunk 13]
```diff
@@ -448,6 +456,24 @@ private extension ManualTrackingViewController {
     }
 }
 
+// MARK: - UISheetPresentationControllerDelegate comformance
+//
+/// Asks permission to dismiss. We call this delegate method whenever the user attempts to dismiss the View Controller via the pull-down gesture.
```
Comment (toupper) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift]: Super not a blocker, I don't think we need to add documentation because it is documented by apple on their sdk.

File: WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift [hunk 14]
```diff
@@ -63,7 +63,8 @@ protocol ManualTrackingViewModel {
 
     var shipmentProvider: ShipmentTrackingProvider? { get set }
     var shipmentProviderGroupName: String? { get set }
-
+    var isEmptyState: Bool { get }
```
Comment (toupper) [inline WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift]: Is this property `isEmptyState` necessary?

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift (hunk 12 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift]: Thanks, I was wondering about this one in case there were extra unaccounted cases, but I see all would fall into the else statement anyway 👍

File: WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift (hunk 6 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift]: Thanks for the tip!

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift (hunk 11 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipment Tracking Section/Add Tracking/ManualTrackingViewController.swift]: At the moment we need to check both to cover the case of "User do not fill any field, but taps dismiss", this would trigger the AlertSheet which we do not want, because there's actually nothing to discard there, can be dismissed directly. This is something that should be fixed on `hasUnsavedChanges` logic though, as you said, if both fields are empty this would return true, which isn't correct. Once this check works as expected, `viewModel.hasUnsavedChanges` should be enough indeed.

File: WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift (hunk 10 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift:236]: That doesn't seem correct indeed, what do you think about something like the following?

```
    var hasUnsavedChanges: Bool {
        
        var returnValue = isEmptyState

        if returnValue {
            // Case isEmptyState = true -> hasUnsavedChanges = false
            returnValue = false
        } else if (providerName.isNilOrEmpty || trackingNumber.isNilOrEmpty) {
            // case providerName or trackingNumber are nilOrEmpty -> hasUnsavedChanges = true
            returnValue = true
        }
        
        return returnValue
    }
```

I was thinking as well an alternative by getting rid of `isEmptyState` and using `canCommit` instead, but this creates a new problem:

```
        var returnValue = canCommit

        if returnValue {
            // Case canCommit = true -> hasUnsavedChanges = false. Problem.
            returnValue = false
        } else if (providerName.isNilOrEmpty || trackingNumber.isNilOrEmpty) {
            // case providerName or trackingNumber are nilOrEmpty -> hasUnsavedChanges = true
            returnValue = true
        }
        
        return returnValue
```
`canCommit = true` and `hasUnsavedChanges = false` are not equiparable in this context, as unless the user taps "Add" changes are not saved and they can still dismiss the data. Maybe create a new check for `hasCommited`/`hasSavedChanges`?

File: WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift (hunk 14 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift:66]: I guess I could get rid of it by using the "not operator" on `canCommit`, right? As is basically the opposite?

**Review by joshheald — APPROVED** [review]
